feat: (platform) Link as a standalone component

DeepakSap14 opened the pull request:
#### Please provide a link to the associated issue.
https://github.com/SAP/fundamental-ngx/issues/1901

#### Please provide a brief summary of this pull request.
- Link implemented as a component with many custom functionalities
- Feature Implementation and Example documentation for Link component.

The design document is here:
https://github.com/SAP/fundamental-ngx/wiki/Platform:-Link-Component-Technical-Design-FRD

If this is a new feature, have you updated the documentation?
This is a new feature; documentation is in same PR.

- [x] Stackblitz works for all examples -- Yes

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/SAP/fundamental-ngx?pullRequest=1903) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/SAP/fundamental-ngx?pullRequest=1903) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/SAP/fundamental-ngx?pullRequest=1903) it.</sub>

InnaAtanasova commented:
Hello @DeepakSap14 , we have a PR opened in Core https://github.com/SAP/fundamental-ngx/pull/1904
Do you think we should merge first the PR in Core and then base your PR on it?

Lokanathan-k commented:
@DeepakSap14  please address the build issues.

fkolar commented:
please check the #1904, to incorporate latest changes and fixes.

fkolar commented on libs/platform/src/lib/components/link/link.component.html:
```diff
@@ -0,0 +1,16 @@
+    (focus)="onFocusEvent()"
+    (focusout)="onFocusOutEvent()"
+>
+    <span class="tooltiptext" *ngIf="toolTipText && tooltipVisibility">{{ toolTipText }}</span>
```
Can you help out with this ?  the toolTipText? What is it good for? Why we can no use native functionality of the achor that will also works with ARIA . 

Please check out ariba-label, label. ariba-title, title. I dont thing we need this. 

Also I dont see icon support here. When we make the component to beinstantiated programatically and placed to the view I woudl expect that I am able to bind all these things. 

the only thing that will stay in the ng-content is actual text, icons positioning it should be something to be handled by component

DeepakSap14 replied:
now implemented toolTip like <a [title]="title text">.

Q. for icon support do i need to put <fd-icon> like below.
<a [title]=".."> 
  <fd-icon [glyph]="icon" [size]="iconSize" *ngIf="icon && iconBefore"></fd-icon>
  <ng-content></ng-content>
  <fd-icon [glyph]="icon" [size]="iconSize" *ngIf="icon && iconAfter"></fd-icon>
</a>

or it will be passed as a content.
<fdp-link [href]=""  [title]="'tooltip text showing on focus'" [disabled]="true">
        <fd-icon [glyph]="'chain-link'" [size]="'l'"></fd-icon>
         Link Text
</fdp-link>.
If it is later case. then how position will get handled by component?

DeepakSap14 replied:
implemented later example show above.

fkolar commented on libs/platform/src/lib/components/link/link.component.html:
```diff
@@ -0,0 +1,16 @@
+    [attr.aria-disabled]="disabled"
+    [ngClass]="{ 'fd-link--emphasized': isEmphasized, 'fd-link--inverted': inverted, 'fdp-link-truncate__txt': !wrap }"
+    (click)="clicked($event)"
+    [ngStyle]="{ width: width }"
```
Why do we need width? fiori spec does not say anything like this. plus there is no truncation.

DeepakSap14 replied:
removed with, width should be coming from parent element.

fkolar commented on libs/platform/src/lib/components/link/link.component.ts:
```diff
@@ -0,0 +1,139 @@
+    changeDetection: ChangeDetectionStrategy.OnPush
+})
+export class LinkComponent implements OnInit, AfterViewInit {
+    private _disabled: boolean = false;
```
move private properties at the bottom ( after public ones)

DeepakSap14 replied:
moved private properties after public ones

fkolar replied:
Deepak , I mean more under, You inputs , outputs are private properties too.

fkolar commented on libs/platform/src/lib/components/link/link.component.ts:
```diff
@@ -0,0 +1,139 @@
+    @ViewChild('link', { static: false }) anchor: ElementRef;
+
+    /** Id for the link */
+    @Input() id?: string;
```
Please try to get aligned. When you start using decorators to be inlined with the property stay with it, or if you prefer to have it on new lines, stay with it but dont mix

DeepakSap14 replied:
now change these into common signature like below
@Input()
id?: string

fkolar commented on libs/platform/src/lib/components/link/link.component.ts:
```diff
@@ -0,0 +1,139 @@
+
+    /** Gives option to truncate content of the link based on width. */
+    @Input()
+    width: string;
```
please see my comments in template

DeepakSap14 replied:
removed width. it is not required.

fkolar commented on libs/platform/src/lib/components/link/link.component.ts:
```diff
@@ -0,0 +1,139 @@
+                this.tooltipVisibility = true;
+                this.cd.detectChanges();
+            }
+        }, timeout);
```
I dont like much this timeout thing for such simple component, everytime you focus we will have to fire CD.. Why we can not stick with native tooltip ?

DeepakSap14 replied:
now tooltip implemented as <a [title]="title text">. so these lines has been removed.

fkolar commented on libs/platform/src/lib/components/link/link.component.ts:
```diff
@@ -0,0 +1,139 @@
+
+    /** Remove Tooltip on focusOut event  */
+    onFocusOutEvent() {
+        this.isfocused = false;
```
please see my comments above.

DeepakSap14 replied:
removed this function.

fkolar commented on libs/platform/src/lib/components/link/link.component.ts:
```diff
@@ -0,0 +1,139 @@
+    /** Throw error for blank text link */
+    ngAfterViewInit() {
+        if (this.anchor.nativeElement.textContent.length === 0) {
+            throw new Error('Mandatory text for fdp-link missing');
```
Do we need to enforce it this way? If you put icon it it can be wihout text, but this functionality will break it.

DeepakSap14 replied:
now i am checking like.
if (this.anchor.nativeElement.innerHtml){}
This will insure that it will work only on Icon as well as only text or both.
if nothing provided then only it will raise error.

JKMarkowski commented:
Hello @DeepakSap14 You can wait actually for https://github.com/SAP/fundamental-ngx/pull/1904 in this PR I provided new link directive, which can be used in your fdp-link.

DeepakSap14 commented:
@JKMarkowski , yes i have to implement using core implementation of the link. i will wait for #1904

DeepakSap14 commented:
@Frank, i am including core link changes into mine. will update here.

fkolar commented:
Any news on this one?

DeepakSap14 commented:
@fkolar , can you review my inline reply on your comments. let me know if my approach is correct on icon and mandatory Link text/icon check.

DeepakSap14 commented:
@fkolar , pushed changes integrated with core/link. please review.

fkolar commented on libs/platform/src/lib/components/link/link.module.ts:
```diff
@@ -0,0 +1,11 @@
+@NgModule({
+    declarations: [LinkComponent],
+    imports: [CommonModule, FundamentalNgxCoreModule],
+    exports: [LinkComponent]
```
Can you double check this line? where you imports whole ngxCore ? you should be fine just importing the link module now?

DeepakSap14 replied:
resolved this. now Importing only LinkModule from FundamentalNgxCoreModule,

DeepakSap14 commented:
pushed changes for review comment on LinkModule import.

DeepakSap14 commented:
@fkolar: , @KevinOkamoto has added more anchor attributes in below Design docs. 
changes Done: 
In template: 
1. changed class to [ngClass]- duplicate class was getting applied.
2. Added hreflang, download, media, rel fields in Link component. 
3. Made DOM light weight by assigning null value wherever required.
eg. when link was not disabled, still in DOM disabled=false was appearing. so made change in template as null value. now these attribute does not appear in DOM.

Design doc:
https://github.com/SAP/fundamental-ngx/wiki/Platform:-Link-Component-Technical-Design

fkolar commented:
Please try to make it mergable so it will not get outdated and you will have to do some additional actions.

DeepakSap14 commented:
@InnaAtanasova , can you please review documentation part of this PR. or assign relevant person someone for this.

fkolar commented:
@DeepakSap14 Can you either close it, do something wiht it so it can be merged?  The issue PR is opened for me 7 days now

DeepakSap14 commented:
will raise separate PR for Documentation. merging this PR.
